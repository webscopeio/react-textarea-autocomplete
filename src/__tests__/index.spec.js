import React from 'react';
import { shallow, mount } from 'enzyme';
import ReactTextareaAutocomplete from './../index';

//eslint-disable-next-line
const SmileItemComponent = ({ entity: { label, text } }) =>
  (<div>
    {' '}{label}{' '}
  </div>);

const Loading = () => <div>Loading...</div>;

const mockedFn = jest.fn();

const rtaComponent = (
  <ReactTextareaAutocomplete
    placeholder={'Write a message.'}
    value={'Controlled text'}
    onChange={mockedFn}
    className={'ownClassName'}
    style={{ background: 'red' }}
    loadingComponent={Loading}
    trigger={{
      ':': {
        output: item => `___${item.text}___`,
        dataProvider: () => [
          { id: 1, label: ':)', text: 'happy_face' },
          { id: 2, label: ':(', text: 'sad_face' },
        ],
        component: SmileItemComponent,
      },
    }}
  />
);
const rta = mount(rtaComponent);

it('match match the snapshot', () => {
  expect(shallow(rtaComponent)).toMatchSnapshot();
});

it('Textarea exists', () => {
  expect(rta.find('textarea')).toHaveLength(1);
});

it('After the trigger was typed, it should appear list of options', async () => {
  rta.find('textarea').simulate('change', { target: { value: 'some test :a' } });
  expect(rta.find('.rta__autocomplete')).toHaveLength(1);
});

it('should display all items', () => {
  expect(rta.find('.rta__autocomplete .rta__list .rta__item')).toHaveLength(2);
});

it('items should be rendered within the component', () => {
  expect(rta.find('.rta__item').first().containsMatchingElement(<SmileItemComponent />)).toEqual(
    true,
  );
});

it('should invoke onchange handler', () => {
  expect(mockedFn).toHaveBeenCalled();
});

it('should close the autocomplete after mouse click', () => {
  const item = rta.find('.rta__item').first();
  item.simulate('click');
  expect(rta.find('.rta__item')).toHaveLength(0);
});

it('text in textarea should be changed', () => {
  expect(rta.find('textarea').node.value).toBe('___happy_face___some test :a');
});
