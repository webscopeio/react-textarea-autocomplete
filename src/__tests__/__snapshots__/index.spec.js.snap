// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match match the snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="rta"
>
    <textarea
        className="rta__textarea ownClassName"
        onChange={[Function]}
        placeholder="Write a message."
        style={
            Object {
                "background": "red",
              }
        }
        value=""
    />
</div>,
  "nodes": Array [
    <div
      className="rta"
>
      <textarea
            className="rta__textarea ownClassName"
            onChange={[Function]}
            placeholder="Write a message."
            style={
                  Object {
                        "background": "red",
                      }
            }
            value=""
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ReactTextareaAutocomplete
        className="ownClassName"
        loadingComponent={[Function]}
        onChange={[Function]}
        placeholder="Write a message."
        style={
                Object {
                        "background": "red",
                      }
        }
        trigger={
                Object {
                        ":": Object {
                          "component": [Function],
                          "dataProvider": [Function],
                          "output": [Function],
                        },
                      }
        }
        value="Controlled text"
/>,
      "_debugID": 5,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ReactTextareaAutocomplete {
        "_reactInternalInstance": [Circular],
        "changeHandler": [Function],
        "cleanUpProps": [Function],
        "closeAutocomplete": [Function],
        "context": Object {},
        "getCurrentTriggerSettings": [Function],
        "getSuggestions": [Function],
        "getTextToReplace": [Function],
        "getValuesFromProvider": [Function],
        "onSelect": [Function],
        "props": Object {
          "className": "ownClassName",
          "loadingComponent": [Function],
          "onChange": [Function],
          "placeholder": "Write a message.",
          "style": Object {
            "background": "red",
          },
          "trigger": Object {
            ":": Object {
              "component": [Function],
              "dataProvider": [Function],
              "output": [Function],
            },
          },
          "value": "Controlled text",
        },
        "refs": Object {},
        "setTextareaCaret": [Function],
        "state": Object {
          "actualToken": "",
          "component": null,
          "currentTrigger": null,
          "data": null,
          "dataLoading": false,
          "left": 0,
          "selectionEnd": 0,
          "selectionStart": 0,
          "top": 0,
          "value": "",
        },
        "textareaRef": null,
        "update": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="rta"
>
          <textarea
                    className="rta__textarea ownClassName"
                    onChange={[Function]}
                    placeholder="Write a message."
                    style={
                              Object {
                                        "background": "red",
                                      }
                    }
                    value=""
          />
</div>,
        "_debugID": 6,
        "_renderedOutput": <div
          className="rta"
>
          <textarea
                    className="rta__textarea ownClassName"
                    onChange={[Function]}
                    placeholder="Write a message."
                    style={
                              Object {
                                        "background": "red",
                                      }
                    }
                    value=""
          />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ReactTextareaAutocomplete
    className="ownClassName"
    loadingComponent={[Function]}
    onChange={[Function]}
    placeholder="Write a message."
    style={
        Object {
            "background": "red",
          }
    }
    trigger={
        Object {
            ":": Object {
              "component": [Function],
              "dataProvider": [Function],
              "output": [Function],
            },
          }
    }
    value="Controlled text"
/>,
}
`;
